### ISSUE TITLE
nep-pack transition to 0.7 (and 1.0)

### ISSUE THREAD

**Issue by jarlebring** [nep-pack/NonlinearEigenproblems.jl#41]
Here is an issue to collect points that need to be addressed when we switch to julia version 0.7. There is an extensive list here: https://github.com/JuliaLang/julia/blob/master/NEWS.md but we can collect the ones relevant for us here:

1.  The syntax for parametric methods, `function f{T}(x::T)`, has been changed to `function f(x::T) where {T}`.
2. `ones(3)+1` will no longer work. Instead one should use `ones(3) .+ 1`. Note it is recommended to use space before the dot, to avoid ambiguity with the dot operator (as in 3 below) 
3. MATLAB-style componentwise version of basic operations is deprecated, `sin(ones(3))` will not work. Use `sin.(ones(3))` instead. (The dot is actually an operation, so you can use it on functions you defined yourself: `f=x-> (x+1)^2;  f.(ones(3))`
4. `eye()` has been deprecated. Instead  use `Matrix(...)`  (not sure what exactly the recommended way is.)
5. `workspace()` is removed (don't know how to do instead). Partial replaced by `revise`?
6.  `qr` methods now return decomposition objects such as QR, QRPivoted, and QRCompactWY rather than tuples of arrays. We need to do a search and replace on all calls to qr.
7.  `expm` is replaced by `exp`. Same for `sqrtm` -> `sqrt`. I fear this will cause very hidden bugs. 
8. All `type` need to be replaced with `struct` or `mutable struct` 
9. `full(A)` should be replaced by `Matrix(A)`
10. "global" definition is necessary to make (some variables) visible inside e.g. a for-loop. See comment below. 

Points 1-3,8,9 can be done without breaking backward compatibility with 0.6, so they are recommended already now.

**Comment by jarlebring** [comment]
The packages we are depending on seem to already have done a transition to 0.7:

* IterativeSolvers: commit https://github.com/JuliaMath/IterativeSolvers.jl/commit/49cc093d60babbf7a3edd0119958d93617771406
* PolynomialRoots: commit: https://github.com/giordano/PolynomialRoots.jl/commit/0db22c566786529febdacd64aac7daa39d33c459
* Polynomials: commit https://github.com/JuliaMath/Polynomials.jl/commit/f81595b42f0db671a40e303c5942e416da2bc98f
* LinearMaps: commit https://github.com/Jutho/LinearMaps.jl/commit/d00a9165bff278029d07f7e88db723c7b433abf6

From that perspective, we are also ready to move to 0.7.

**Comment by eringh** [comment]
I was thinking about whether we should go via 0.7 or directly to 1.0. Hence for our own future reference:
0.7 and 1.0 are very similar. If it runs in 0.7 without warnings it should work on 1.0. Therefore the recommended development strategy is to upgrade to 0.7 first and then to 1.0.

See: [https://discourse.julialang.org/t/what-is-julia-0-7-how-does-it-relate-to-1-0/9994](https://discourse.julialang.org/t/what-is-julia-0-7-how-does-it-relate-to-1-0/9994)

**Comment by jarlebring** [comment]
Richard Tsai pointed out there is a difference in scope of variables in 0.7. For instance, inside a for-loop an int defined outside is in general not visible:
```
julia> x=0;
or
julia> for i=1:10
       x+=1;
       end
┌ Warning: Deprecated syntax `implicit assignment to global variable `x``.
│ Use `global x` instead.
└ @ none:0
```
A bit counter intuitive, no "global" is necessary if its an array:
```
julia> x=[0];
julia> for i=1:10
       x[1]+=1;
       end
```

**Comment by maxbennedich** [comment]
Good to know. My understanding is that the scope change is only a problem if the parent scope is global, such as the REPL. Running that code from within a function for example works as expected, so hopefully it won't introduce any issues in our package.

**Comment by maxbennedich** [comment]
I've started working on the migration to Julia 0.7 in the `v0.7` branch. In order to reduce the workload and risk of keeping that branch in sync with master, we could do a release of the current master branch (which would be a Julia 0.6 version), and then merge the `v0.7` branch to master. It would mean that master would not build for a period until we've finished the upgrade, but hopefully we'll be able to get the majority done within a week or so. (If needed, we could still branch off of an older working version to do work in Julia 0.6.)

I've spoken to @eringh about his 'jd' branch, and he believes that it's better to hold off and merge that to master once we've done the v0.7 upgrade. @meleg also has nothing outstanding at the moment for such a release.

Anything else that someone wants to get into a v0.6 release, or are we good to go?

**Comment by jarlebring** [comment]
I have nothing that I need to get into that release. I think @upadpup's work on nlarnoldi made it into a stable shape so it can be on hold for a week or so. Do you agree @upadpup?

**Comment by upadpup** [comment]
I would say most of the short-term ideas with nlarnoldi have been implemented(more or less). For the more long-term goals(like restarting), I will need some time to read up some papers anyway. So holding off for a week is fine by me.

**Comment by maxbennedich** [comment]
Great, then we could do the release now, after which I'll merge the v0.7 branch into master. @jarlebring would you like to create the release?

**Comment by jarlebring** [comment]
Done. I was not able to make an installation of a specific tag in a one-liner. If someone figures it out, you're welcome to update the README.md.

**Comment by maxbennedich** [comment]
I've merged the v0.7 branch into master now, which means that master will be broken for a while until we fix all the incompatibilities. I'll try to put instructions on how to get started on v0.7 on [the wiki](https://github.com/nep-pack/NonlinearEigenproblems.jl/wiki/Julia-v0.7-transition) over the next couple of days.

**Comment by maxbennedich** [comment]
There is a change in behavior in `indmin` and `indmax` in terms of how they handle `NaN` values. In Julia 0.6:
```
julia> indmin([NaN, 3])
2
```
In Julia 0.7:
```
julia> indmin([NaN, 3])
1
```
(Also note that the methods have been deprecated and replaced by `argmin` and `argmax`.)

In NLEIGS, this change in behavior introduced a silent bug. `indmin` is also used in `method_broyden.jl`, `Gallery_old.jl` and `NEPSolver.jl`. Does anyone know if `NaN` values occur in these instances?

**Comment by jarlebring** [comment]
Very good to know. I think `NaN` in an eigenvalue list in broyden would mean that the method failed. It does not seem so dangerous there.

**Comment by maxbennedich** [comment]
Ok good. Here's a concise workaround just in case:

`indmin(v)` => `argmin([isnan(x) ? Inf : x for x in v])`

If this hurts performance, perhaps explicit iteration (e.g. a for loop) is a better choice.

**Comment by maxbennedich** [comment]
The NLEIGS tests are now running successfully in Julia 0.7, and are around 2 - 28 % faster than in Julia 0.6.4, depending on the problem. Details below. (Each test has been run 10+ times, and the fastest run is reported.)
```
gun_variant_s
MATLAB        5.802198 seconds
Julia 0.6.4   4.897908 seconds (2.93 M allocations: 2.318 GiB, 8.97% gc time)
Julia 0.7.0   4.795207 seconds (1.57 M allocations: 2.235 GiB, 9.24% gc time)

gun_variant_r2
MATLAB        23.804606 seconds
Julia 0.6.4   15.485986 seconds (2.94 M allocations: 6.837 GiB, 6.05% gc time)
Julia 0.7.0   15.202975 seconds (1.58 M allocations: 6.754 GiB, 5.19% gc time)

particle_variant_s
MATLAB        16.091790 seconds
Julia 0.6.4   9.681309 seconds (32.06 M allocations: 8.739 GiB, 20.11% gc time)
Julia 0.7.0   7.879619 seconds (8.77 M allocations: 8.096 GiB, 20.27% gc time)

particle_variant_r2
MATLAB        18.215674 seconds
Julia 0.6.4   13.136384 seconds (32.21 M allocations: 7.424 GiB, 28.37% gc time)
Julia 0.7.0   9.465336 seconds (8.89 M allocations: 6.778 GiB, 14.31% gc time)

Total MATLAB        63.9 s
Total Julia 0.6.4   43.2 s
Total Julia 0.7.0   37.3 s
```

**Comment by jarlebring** [comment]
Awesome. And awesome work.

**Comment by maxbennedich** [comment]
In Julia 0.6:
```
julia> norm([1 0; 0 1])
1.0
```
In Julia 0.7:
```
julia> norm([1 0; 0 1])
1.4142135623730951
```
I.e., `norm` used to return the matrix norm for matrices, but now returns the Frobenius norm. To get the matrix norm, there's a new function called `opnorm`. This will most likely introduce errors in our package. I've found and [fixed one already](https://github.com/nep-pack/NonlinearEigenproblems.jl/commit/47dca66334620494fa9d4d8f461e2c526ce055d6).

We are calling `norm` 334 times in our package, and it's not always clear when the argument is a matrix. Should all of the matrix instances be replaced by `opnorm`, or are there cases where it doesn't matter, or where the Frobenius norm is more suitable?

One way to find when this is called with a matrix argument, is to replace _all_ occurrences of `norm` with `opnorm`, which will throw an error when called with a vector, and then go through and fix all the errors... I think this might be the safest option, since it will ensure the same behavior as today.

**Comment by jarlebring** [comment]
That's a very tricky change (and not convincing justification in my opinon, cf julialang PR 27401 ). 

One thing we anyway may want to do would lessen this issue:  I would suggest we first change all occurrances of  `v=v/norm(v)` to `normalize!(v)`, and `u=v/norm(v)` by `u=normalize(v)`.  That should reduce the number of cases that we have to look carefully at. 

Heads up for tiar: It is sensitive to change of norm in the coefficient computation.

**Comment by maxbennedich** [comment]
Indeed not convincing! So for now I did a search and replace for `norm` -> `opnorm` and then repeatedly ran all tests and fixed the failures where `opnorm` was called with a vector. This means that all tested code should work, but there are likely several untested code paths which will throw an error when executed. I think this is safer than silently returning the incorrect norm for matrices, and manually inspecting hundreds of calls to figure out the argument type was too overwhelming for me, especially since I'm not so familiar with much of the code. [The diff is here](https://github.com/nep-pack/NonlinearEigenproblems.jl/commit/e5fe4099fe599f3c3149c791cf136f45f4688beb) if someone wants to take a look and change any obvious errors where `opnorm` is called for vectors back to `norm`.

**Comment by jarlebring** [comment]
The `qr` function for rectangular matrices has changed:

julia 0.6:
```
julia> Q0,R0=qr(randn(5,3)); size(Q0)
(5, 3)
```
julia 0.7:
```
julia> Q0,R0=qr(randn(5,3)); size(Q0)
(5, 5)
```
That is, it does not compute the *thin* qr-factorization for rectangular matrices. 

It needs to be done by returning a struct:
```
julia> F=qr(randn(5,3));
julia> Q1=Matrix(F.Q); size(Q1)
(5, 3)
```

Note: Doing `Matrix(Q0)` will still lead to the matrix 5x5 (!?). Very confusing.

**Comment by maxbennedich** [comment]
> The qr function for rectangular matrices has changed:

Yes I noticed that too. It should have been taken care of in [this commit](https://github.com/nep-pack/NonlinearEigenproblems.jl/commit/3c3964e66ca66bf0efa1b7f7bf6a6a2f19f55176). The recommended solution seemed to be to just call Matrix(Q)?

**Comment by jarlebring** [comment]
Yes. "Note" in my post was incorrect.

**Comment by maxbennedich** [comment]
![image](https://user-images.githubusercontent.com/1812661/44952642-ab709b80-ae84-11e8-94e6-7cffceb9df45.png)

**Comment by maxbennedich** [comment]
Code coverage is unfortunately not accurate at all anymore. I've tried disabling inlining, but that didn't make any difference. This needs to be investigate further.

**Comment by maxbennedich** [comment]
Although all tests now build and pass, some are considerably slower and use more memory than before. See the comparison below, where the first line for each test is the result in Travis for a 0.6 build, and the second line is a 0.7 build. Keep in mind that measurements usually vary quite a bit between runs, so some fluctuation would be expected, but several tests have obviously regressed, especially where both elapsed time and memory have increased. I think these should be analyzed in more detail. Perhaps some of these are related to the transpose issue discussed in #58.
```
Julia 0.6 total measured:  452s  76.9GiB
Julia 0.7 total measured:  545s  93.8GiB

 Section                             ncalls     time   %tot     alloc   %tot
 ───────────────────────────────────────────────────────────────────────────
 beyn                                     1    13.7s  3.03%    455MiB  0.58%
 beyn                                     1    57.3s  10.5%   1.56GiB  1.66%

 nlar                                     1    52.1s  11.6%   11.5GiB  15.0%
 nlar                                     1    49.4s  9.09%   11.2GiB  12.0%
 
 iar_chebyshev                            1    37.9s  8.41%   3.51GiB  4.56%
 iar_chebyshev                            1    42.3s  7.78%   4.91GiB  5.25%
 
 wep_small                                1    34.4s  7.63%   8.89GiB  11.6%
 wep_small                                1    40.6s  7.46%   9.51GiB  10.2%

 nleigs/nleigs_gun_variant_r2             1    28.9s  6.40%   7.09GiB  9.23%
 nleigs/nleigs_gun_variant_r2             1    28.7s  5.28%   6.96GiB  7.43%
 
 bigfloats                                1    19.3s  4.27%    474MiB  0.60%
 bigfloats                                1    24.8s  4.56%   2.12GiB  2.27%
 
 nleigs/nleigs_particle_variant_r2        1    24.6s  5.46%   7.84GiB  10.2%
 nleigs/nleigs_particle_variant_r2        1    23.2s  4.28%   7.38GiB  7.88%
 
 compute_eigvec_from_eigval_lu            1    22.5s  5.00%   3.41GiB  4.44%
 compute_eigvec_from_eigval_lu            1    23.2s  4.26%   3.58GiB  3.82%
 
 compute_eigvec_from_eigval_lopcg         1    13.4s  2.96%   1.17GiB  1.52%
 compute_eigvec_from_eigval_lopcg         1    23.1s  4.24%   6.06GiB  6.47%

 compan                                   1    13.1s  2.90%    483MiB  0.61%
 compan                                   1    22.2s  4.08%   2.14GiB  2.29%
 
 nleigs/nleigs_particle_variant_s         1    18.7s  4.15%   9.13GiB  11.9%
 nleigs/nleigs_particle_variant_s         1    16.8s  3.10%   8.55GiB  9.13%
 
 newton_test                              1    7.44s  1.65%    908MiB  1.15%
 newton_test                              1    16.4s  3.01%   2.12GiB  2.26%
 
 nleigs/nleigs_gun_variant_p              1    14.2s  3.15%   2.15GiB  2.79%
 nleigs/nleigs_gun_variant_p              1    15.8s  2.91%   2.35GiB  2.51%
 
 jd                                       1    15.7s  3.49%    569MiB  0.72%
 jd                                       1    15.6s  2.87%   1.41GiB  1.51%
 
 nleigs/nleigs_basic                      1    21.3s  4.73%   1.70GiB  2.21%
 nleigs/nleigs_basic                      1    13.3s  2.45%   1.28GiB  1.36%
 
 broyden                                  1    13.0s  2.87%    586MiB  0.74%
 broyden                                  1    12.8s  2.35%   1.43GiB  1.52%

 infbilanczos                             1    10.2s  2.26%   4.44GiB  5.77%
 infbilanczos                             1    11.7s  2.16%   4.53GiB  4.84%

 nleigs/nleigs_gun_variant_r1             1    11.2s  2.48%   2.15GiB  2.79%
 nleigs/nleigs_gun_variant_r1             1    11.0s  2.02%   2.05GiB  2.19%

 nleigs/nleigs_gun_variant_s              1    10.2s  2.27%   2.55GiB  3.32%
 nleigs/nleigs_gun_variant_s              1    9.93s  1.83%   2.42GiB  2.59%

 nleigs/nleigs_nep_types                  1    7.85s  1.74%    922MiB  1.17%
 nleigs/nleigs_nep_types                  1    9.01s  1.66%   1.09GiB  1.16%

 transf                                   1    8.76s  1.94%   1.07GiB  1.39%
 transf                                   1    8.46s  1.56%   1.18GiB  1.26%

 tiar                                     1    11.8s  2.62%   2.80GiB  3.64%
 tiar                                     1    8.42s  1.55%   2.65GiB  2.83%

 periodicdde                              1    3.50s  0.78%    755MiB  0.96%
 periodicdde                              1    7.95s  1.46%   1.30GiB  1.39%

 inner_solves                             1    3.28s  0.73%    164MiB  0.21%
 inner_solves                             1    6.27s  1.15%    723MiB  0.75%

 blocknewton                              1    4.41s  0.98%    179MiB  0.23%
 blocknewton                              1    5.50s  1.01%    485MiB  0.51%

 dep_distributed                          1    5.11s  1.13%    331MiB  0.42%
 dep_distributed                          1    5.47s  1.01%    595MiB  0.62%

 deflation                                1    4.34s  0.96%    117MiB  0.15%
 deflation                                1    5.10s  0.94%    370MiB  0.39%

 iar                                      1    3.16s  0.70%    110MiB  0.14%
 iar                                      1    4.14s  0.76%    300MiB  0.31%

 gun_native                               1    4.05s  0.90%    546MiB  0.69%
 gun_native                               1    4.00s  0.74%    569MiB  0.59%

 spmf                                     1    3.62s  0.80%   68.3MiB  0.09%
 spmf                                     1    3.73s  0.69%    225MiB  0.23%

 proj                                     1    2.80s  0.62%    138MiB  0.18%
 proj                                     1    3.26s  0.60%    467MiB  0.49%

 nleigs/discretizepolygon                 1    1.48s  0.33%   27.4MiB  0.03%
 nleigs/discretizepolygon                 1    2.09s  0.38%    151MiB  0.16%

 nleigs/nleigs_scalar                     1    1.48s  0.33%    746MiB  0.95%
 nleigs/nleigs_scalar                     1    1.82s  0.34%    682MiB  0.71%

 sgiter                                   1    1.18s  0.26%   33.5MiB  0.04%
 sgiter                                   1    1.38s  0.25%    150MiB  0.16%

 core                                     1    831ms  0.18%   17.5MiB  0.02%
 core                                     1    1.20s  0.22%   72.2MiB  0.08%

 mslp                                     1    361ms  0.08%   5.76MiB  0.01%
 mslp                                     1    844ms  0.16%   51.3MiB  0.05%

 nleigs/cached_lu_solver                  1    536ms  0.12%   10.0MiB  0.01%
 nleigs/cached_lu_solver                  1    773ms  0.14%   60.0MiB  0.06%

 nleigs/inpolygon                         1    370ms  0.08%   8.90MiB  0.01%
 nleigs/inpolygon                         1    421ms  0.08%   33.0MiB  0.03%

 serialization                            1    164ms  0.04%   1.95MiB  0.00%
 serialization                            1    393ms  0.07%   21.8MiB  0.02%
```

**Comment by jarlebring** [comment]
The fact that beyn seem to have regressed might to have something to do with the fact that it is the first test to run. I permuted the list, and CPU and memory both improved so it was further down the list. Something with loading/compiling the packages maybe.

**Comment by maxbennedich** [comment]
Ah, good point.

I noticed something similar for the `inpolygon` and `discretizepolygon` tests. They don't really allocate much memory at all, but since the tests are only run and timed once, the measurements include JIT compilation, which seems to use a lot more memory in Julia 0.7 than in 0.6. So the numbers above are even less reliable than I thought.

**Comment by maxbennedich** [comment]
I've enabled Travis builds for Julia 1.0 in addition to 0.7. Only minimal changes were required to make our code (or rather, what's tested of it) compatible with 1.0.

It may appear in Travis as if the build takes twice as long, however the two builds (for 0.7 and 1.0) are done in parallel, so the wall clock build time is about the same as before.

**Comment by maxbennedich** [comment]
Below is an updated list of tests that either have gotten slower, or allocate more memory, after the 0.7 upgrade. This time I've run each test twice in a row, and only measure the time of the second run. This vastly reduces the impact of JIT compilation on the numbers (but doesn't eliminate it completely).

There are three measurements for each test: The first line is for the last Julia 0.6 version before the upgrade, the second line is for the Julia 0.7 version right after the upgrade, and the third line is the current version run in Julia 1.0.

Several of these are likely false positive due to some JIT compilation time still being included, but at least some are probably genuine regressions, such as `iar_chebyshev` (I'll look at that one first).
```
v0.6, v0.7, v1.0                  ncalls     time   %tot     alloc   %tot
───────────────────────────────────────────────────────────────────────────
nlar                                   1    30.5s  18.5%   12.2GiB  17.7%
nlar                                   1    33.2s  19.0%   11.5GiB  15.6%
nlar                                   1    34.0s  19.1%   11.5GiB  15.5%

iar_chebyshev                          1    392ms  0.24%    144MiB  0.20%
iar_chebyshev                          1    9.68s  5.56%   3.10GiB  4.20%
iar_chebyshev                          1    10.9s  6.12%   3.13GiB  4.23%

newton_test                            1    2.54s  1.54%    767MiB  1.09%
newton_test                            1    5.90s  3.38%   1.37GiB  1.85%
newton_test                            1    6.00s  3.38%   1.36GiB  1.84%

compute_eigvec_from_eigval_lopcg       1    1.32s  0.80%    880MiB  1.25%
compute_eigvec_from_eigval_lopcg       1    4.68s  2.69%   4.78GiB  6.48%
compute_eigvec_from_eigval_lopcg       1    3.07s  1.73%   4.76GiB  6.43%

periodicdde                            1    2.06s  1.25%    736MiB  1.05%
periodicdde                            1    3.05s  1.75%    912MiB  1.21%
periodicdde                            1    2.95s  1.66%    911MiB  1.20%

nleigs/nleigs_nep_types                1    2.65s  1.61%    311MiB  0.44%
nleigs/nleigs_nep_types                1    2.90s  1.67%    652MiB  0.86%
nleigs/nleigs_nep_types                1    2.92s  1.65%    649MiB  0.86%

dep_distributed                        1    1.25s  0.76%    244MiB  0.35%
dep_distributed                        1    1.42s  0.81%    266MiB  0.35%
dep_distributed                        1    1.40s  0.79%    264MiB  0.35%

iar                                    1    849ms  0.52%   81.9MiB  0.12%
iar                                    1    999ms  0.57%    126MiB  0.17%
iar                                    1    974ms  0.55%    131MiB  0.17%

bigfloats                              1    494ms  0.30%   10.1MiB  0.01%
bigfloats                              1    660ms  0.38%   76.9MiB  0.10%
bigfloats                              1    808ms  0.46%   78.0MiB  0.10%

nleigs/discretizepolygon               1    689ms  0.42%   15.0MiB  0.02%
nleigs/discretizepolygon               1    680ms  0.39%   34.1MiB  0.05%
nleigs/discretizepolygon               1    713ms  0.40%   33.7MiB  0.04%

proj                                   1    388ms  0.24%   62.6MiB  0.09%
proj                                   1    456ms  0.26%   69.5MiB  0.09%
proj                                   1    511ms  0.29%   67.1MiB  0.09%

spmf                                   1    337ms  0.20%   5.21MiB  0.01%
spmf                                   1    443ms  0.25%   12.5MiB  0.02%
spmf                                   1    414ms  0.23%   10.4MiB  0.01%

inner_solves                           1    312ms  0.19%    108MiB  0.15%
inner_solves                           1    341ms  0.20%    116MiB  0.15%
inner_solves                           1    352ms  0.20%    115MiB  0.15%

nleigs/inpolygon                       1    161ms  0.10%   4.17MiB  0.01%
nleigs/inpolygon                       1    256ms  0.15%   21.6MiB  0.03%
nleigs/inpolygon                       1    260ms  0.15%   20.6MiB  0.03%

nleigs/cached_lu_solver                1    149ms  0.09%   2.89MiB  0.00%
nleigs/cached_lu_solver                1    236ms  0.14%   15.9MiB  0.02%
nleigs/cached_lu_solver                1    227ms  0.13%   14.1MiB  0.02%
```

**Comment by jarlebring** [comment]
Interesting! The newton test has changed so I'm sure that is worth investigating. cached_lu_solver is such a small time so it can be within the noise level.

**Comment by maxbennedich** [comment]
Another explanation for some tests behaving differently in v0.7 is that randomly generated numbers (which are used as start vectors for some tests) could differ. This explained most of the difference for `compute_eigvec_from_eigval_lopcg`.

**Comment by maxbennedich** [comment]
As discussed in #63, we will now drop support for Julia 0.7. (Which just means updating Travis to not build 0.7, and re-adding PolynomialZeros which doesn't work properly in 0.7.)

@jarlebring @eringh @meleg @upadpup: You should probably upgrade your Julia installation to 1.0, if you haven't already.

**Comment by maxbennedich** [comment]
I've found another unpleasant regression in Julia 0.7 / 1.0: matrix exponential can be a lot slower, at least for small matrices. This is what caused the regression for at least `periodicdde` and `dep_distributed ` in the table above, and probably affects more methods since we seem to use matrix exponential quite a bit.

For example, in Julia 0.6.4:
```julia
julia> @btime expm([1 0; 2 0])
  3.601 μs (46 allocations: 4.55 KiB)
```
In Julia 0.7 and 1.0:
```julia
julia> @btime exp([1 0; 2 0])
  16.710 μs (67 allocations: 4.55 KiB)
```

When the matrix size grows, the difference seems to be smaller.

Does anyone know if this will be an issue for actual applications of our package, or is it only in our unit tests that we're doing exponential with such small matrices (let's say smaller than 40x40)?

I've filed a github issue with Julia to look into this.

**Comment by maxbennedich** [comment]
Alright, I've now gone through all tests that showed a regression in performance between 0.6.4 and 0.7, and the majority were caused by differences in JIT compilation. Note that it's not enough to `include` a test twice to fully remove JIT compilation from the measurements, it needs to be done on a more granular level such as in my suggestion in #67.

There are two tests that are still slower (`periodicdde` and `dep_distributed`) due to an issue I found with matrix exponential within Julia. After I reported the issue, it only took half an hour for someone to fix it, and it seems that they'll try to backport this fix to Julia 1.0.

The remaining tests are either as fast as before, or more commonly a bit faster.

While there will surely be some remaining work related to this transition (especially in untested/unreachable code -- we still have references to `eye` in our code, which will fail if run), I'm starting to feel that the activate transition phase is completing, and that we can deal with remaining issues as they come up.
